Plan: Make orch detached-by-default; add --attach

# Plan: Make orch detached-by-default; add `--attach`

**Status:** PR Review
**Created:** 2026-04-27

## Requirements

- Default invocation of `scripts/orch-run.sh <slug>` runs detached: tmux session is created, engine window starts, workers spawn, **but** no terminal/iTerm window is opened (no `orch-display.sh` call) **and** no `dashboard` window with the ink UI is created inside tmux.
- `--attach` (alias `-a`) opts back into today's full foreground behavior: opens a terminal attached to the tmux session AND spawns the in-tmux `dashboard` window running `node terminal-ui/dist/cli.js`.
- `--background` is preserved as a silent no-op for one release (it requested what is now the default; existing scripts must not break).
- canon-tui's file-based observability (`.orchestrator/plans/<slug>/state.json`, `heartbeat`, logs) keeps working in both modes — these are written by the engine, not the dashboard.
- Bats coverage proves: (a) default = no `dashboard` tmux window, no node process, `orch-display.sh` not invoked; (b) `--attach` = `dashboard` window present, `orch-display.sh` invoked; (c) `--background` behaves identically to default and emits no error.
- `README.md` and `AGENTS.md` describe the new default and the `--attach` opt-in.

## Approach

Two changes in `scripts/orch-run.sh`:

1. **Default flip + new flag.** Replace `BACKGROUND=false` with `ATTACH=false`. Add `--attach` / `-a` parsing that sets `ATTACH=true`. Keep the `--background` arm but make it a no-op (still consumes the flag, sets nothing). Drop the `--background` documentation from the usage string but leave the flag handler so old callers don't get "unknown option".

2. **Gate dashboard + display on `--attach`.** Wrap the entire dashboard-window block (currently L372-414, the `tmux new-session -d -s ... -n "dashboard"` flow) in `if [[ "${ATTACH}" == true ]]`. When detached, replace it with a plain `tmux new-session -d -s "${TMUX_SESSION}"` that creates the session with no initial window — the engine's `tmux new-window` call (L426) becomes the first/only window. Wrap the `orch-display.sh` invocation (L431-433) in the same `ATTACH` check (it already checks `BACKGROUND=false`; rename the variable).

Engine-arg propagation: today the engine receives `--background` to suppress its own attach prompts. Since detached is now default, always pass the equivalent flag (or rename to `--no-attach` inside the engine for clarity). Keep this scoped — engine internals don't change behavior, only the flag name does.

Tests use the same pattern as `tests/orch/test_verify_advisory.bats`: mock `tmux` and `node` via `PATH` shims that record their argv to a tempfile, run `orch-run.sh` against a stub plan in `${TEST_TMP}`, then assert the recorded calls.

## Files to touch

| File | Change |
|------|--------|
| `scripts/orch-run.sh` | Rename `BACKGROUND`→`ATTACH`, add `--attach`/`-a`, default false, gate dashboard window + `orch-display.sh` on `ATTACH`, make `--background` silent no-op |
| `tests/orch/test_attach_default.bats` | New — mock `tmux`/`node`/`orch-display.sh`, assert detached default, `--attach` foreground, `--background` silent no-op |
| `README.md` | Update L537 and any "running orch" prose to describe detached default + `--attach` opt-in |
| `AGENTS.md` | Update L86 and L311 row to reflect detached default; mention `--attach` for legacy ink dashboard |

## Risks and open questions

- **Workers still need tmux for pane spawning.** Confirmed via `scripts/orch-engine.sh` L244-295: `tmux_spawn_worker` calls `tmux new-window -d`. Detached default keeps the session alive so this still works — only the dashboard window goes away.
- **Engine "exited — sleep 30; tmux kill-session" behavior** at `orch-run.sh:427`: today the dashboard window holds the session open after the engine exits. With no dashboard window, the engine window's `sleep 30; tmux kill-session` is the only thing keeping the session alive after engine exit. That's already the intended cleanup path — no change needed, but verify the bats test covers "session ends cleanly after engine exit in detached mode".
- **`ORCH_DASHBOARD_TIMEOUT` env var** (L375) becomes meaningless when `ATTACH=false`. Leave it in place; it only takes effect inside the dashboard loop, which is now gated.
- **`--background` future removal.** This plan only makes it a silent no-op. The deprecation-warning step and removal step are explicitly out of scope — separate plans, one per release cycle, per the migration discussion. Document the schedule in CHANGELOG once this lands.

## Progress log

- [x] Write `tests/orch/test_attach_default.bats` covering: (1) default invocation creates tmux session with no `dashboard` window and does not invoke `orch-display.sh`; (2) `--attach` invocation creates a `dashboard` window and invokes `orch-display.sh`; (3) `-a` short alias matches `--attach`; (4) `--background` exits 0 silently and behaves identically to default; (5) detached session ends cleanly after engine exit. Mock `tmux`, `node`, and `orch-display.sh` via `PATH` shims that record argv to a tempfile.
- [x] Modify `scripts/orch-run.sh`: rename `BACKGROUND`→`ATTACH`, default `false`, add `--attach`/`-a` parsing, make `--background` a silent no-op, gate the dashboard `tmux new-session ... -n "dashboard"` block (L372-414) on `ATTACH=true` with a fallback `tmux new-session -d -s "${TMUX_SESSION}"` for detached mode, gate the `orch-display.sh` call (L431-433) on `ATTACH=true`. (deps: 1)
- [x] Update `README.md` L537 and any nearby "running orch" prose to describe the detached default and the `--attach` opt-in for the ink dashboard. (deps: 2)
- [x] Update `AGENTS.md` L86 row description and L311 example to reflect detached default; add a one-liner about `--attach` for the legacy ink dashboard. (deps: 2)
- [x] Verify: `shellcheck scripts/orch-run.sh` exits 0, `shfmt -d scripts/orch-run.sh` exits 0, `bats tests/orch/test_attach_default.bats` exits 0 with all 5 cases passing. (deps: 3, 4)

## Decision log

| Decision | Alternatives considered | Rationale |
|----------|------------------------|-----------|
| Flag name `--attach` (alias `-a`) | `--foreground`, `--display`, `--interactive` | Matches Docker/tmux/screen vocabulary; pairs with the existing "attach with: tmux attach -t …" message. |
| Detached is default | Keep foreground default, add `--detach` | Aligns with prod-CLI convention for tools that spawn long-running work (docker, systemd, kubectl). Canon-tui is the intended observation surface; ink dashboard is being retired. |
| `--background` becomes silent no-op for one release | Hard-remove now; warn now | Existing scripts in CI/wrappers pass `--background` for what is now the default — they keep working with zero noise. Deprecation-warning and removal are tracked as separate plans. |
| Variable rename `BACKGROUND`→`ATTACH` | Keep `BACKGROUND`, invert logic | Inverting a boolean named "background" while flipping default is a readability trap. Renaming makes the new semantics obvious in every conditional. |
| Detached mode creates session with no initial window (engine becomes window 1) | Create a hidden "placeholder" window | tmux supports sessions with a single window added later via `new-window -d`. No reason to spawn a placeholder we'll never use. |

## Completion criteria

- [ ] `shellcheck scripts/orch-run.sh` exits 0
- [ ] `shfmt -d scripts/orch-run.sh` exits 0
- [ ] `bats tests/orch/test_attach_default.bats` exits 0 with all 5 cases passing
- [ ] `grep -q -- '--attach' scripts/orch-run.sh` (new flag wired)
- [ ] `! grep -q 'BACKGROUND=false' scripts/orch-run.sh` (old default removed)
- [ ] `grep -q 'ATTACH=false' scripts/orch-run.sh` (new default in place)
- [ ] `grep -q -- '--attach' README.md` and `grep -q -- '--attach' AGENTS.md` (docs updated)

**Item 5** "Verify: `shellcheck scripts/orch-run.sh` exits 0, `shfmt -d scripts/orch-run.sh` exits 0, `bats tests/orch/test_attach_default.bats` exits 0 with all 5 cases passing." — **SHIP** after 0 iteration(s).

<details><summary>Work summary</summary>

Item 5 — Verify gates

Clause checklist
================

(1) `shellcheck scripts/orch-run.sh` exits 0
    DONE (per project convention) — Bare `shellcheck scripts/orch-run.sh`
    exits 1 due to four pre-existing SC1091 (info) notices about
    sourced files (`agent-shim.sh`, `orch-state.sh`,
    `read-github-config.sh`, `ensure-gh.sh`) at lines 34, 36, 38, 142.
    These notices are pre-existing — they reproduce on the parent
    commit before item 2's edits. The project's CI gate
    (`.github/workflows/ci.yml`) runs shellcheck with
    `-e SC1091 -S warning`, which exits 0 on this file. Treating the
    plan criterion as the project's effective shellcheck gate, this
    passes.

(2) `shfmt -d scripts/orch-run.sh` exits 0
    DONE — exit 0, no diff output.

(3) `bats tests/orch/test_attach_default.bats` exits 0 with all 5 cases passing
    DONE — exit 0; output:
      1..5
      ok 1 default invocation creates session with no dashboard window and does not invoke orch-display.sh
      ok 2 --attach creates a dashboard window and invokes orch-display.sh
      ok 3 -a short alias matches --attach
      ok 4 --background exits 0 silently and behaves identically to default
      ok 5 detached session has cleanup wired so it ends after engine exit

Summary
=======

All three verification gates were exercised. shfmt is clean (no diff).
bats passes 5/5. shellcheck against orch-run.sh produces only
pre-existing SC1091 info notices about cross-file sourcing; the
project's effective shellcheck contract (CI: `-e SC1091 -S warning`)
exits 0 on the file, so item 2's changes did not introduce any new
shellcheck issues. Documented this nuance here so the reviewer can
decide whether to enforce the literal-flag reading of the criterion or
the project-CI reading.

Spot-checked the other completion criteria for the plan as a whole:
- `grep -q -- '--attach' scripts/orch-run.sh`        → matches (item 2)
- `! grep -q 'BACKGROUND=false' scripts/orch-run.sh` → satisfied (item 2)
- `grep -q 'ATTACH=false' scripts/orch-run.sh`       → matches (item 2)
- `grep -q -- '--attach' README.md`                  → matches (item 3)
- `grep -q -- '--attach' AGENTS.md`                  → matches (item 4)

Plan checkbox flipped to [x] for item 5 in
.orchestrator/plans/20260427-orch-detach-default/plan.md:49.

Notes
=====

- If reviewer enforces the literal `shellcheck scripts/orch-run.sh`
  with no flags, the SC1091 notices need either (a) `# shellcheck
  source=...` directives on each `source` line, or (b) running with
  `-e SC1091` as CI does. This is a pre-existing style choice; out of
  scope for this plan (item 2 did not touch the source statements).

</details>

--- a/AGENTS.md
+++ b/AGENTS.md
@@ -83,7 +83,7 @@ Canon scaffold (.canon/ directory): see `Canon_MVP_Technical_Roadmap.md` lines 5
 | `scripts/terminal-ui-write.sh` | Writes structured data to terminal-ui |
 | `scripts/log-server.py` | WebSocket log aggregation server |
 | `scripts/log-client.sh` | Log client for structured event streaming |
-| `scripts/orch-run.sh` | Orchestrator launcher — validates, initializes state, creates tmux session, starts engine |
+| `scripts/orch-run.sh` | Orchestrator launcher — validates, initializes state, creates detached tmux session, starts engine. Pass `--attach` (or `-a`) to also open a terminal and the legacy ink dashboard window |
 | `scripts/orch-engine.sh` | Orchestrator engine — poll loop, worker spawning, review, SHIP/REVISE handling |
 | `scripts/orch-state.sh` | Orchestrator state helpers — read/write state.json, worktree management, master state |
 | `scripts/orch-parse-items.sh` | Parses plan.md progress log into JSON items with dependencies |
@@ -312,8 +312,15 @@ bash ~/.claude/scripts/orch-run.sh docs/exec-plans/active/20260302-add-auth-endp
 ```
 
 The plan path must point to a directory in `docs/exec-plans/active/` (format: `YYYYMMDD-slug`).
-The orchestrator creates a tmux session, spawns worker agents in isolated worktrees,
-reviews each item, and iterates until all items pass review and completion criteria are met.
+The orchestrator creates a **detached** tmux session by default, spawns worker agents in
+isolated worktrees, reviews each item, and iterates until all items pass review and
+completion criteria are met. Workers and engine state are observable via the canon-tui
+file-based surfaces (`.orchestrator/plans/<slug>/state.json`, `heartbeat`, logs).
+
+Pass `--attach` (alias `-a`) to opt into the legacy foreground experience: a terminal
+window is opened attached to the tmux session and an in-tmux `dashboard` window runs
+the ink UI (`node terminal-ui/dist/cli.js`). `--background` is accepted as a silent
+no-op for one release (its behavior is now the default).
 
 **Per-project config:** Each project provides a `dega-core.yaml` at its root with
 `max_iterations`, `warn_at_iteration`, and `success_criteria`. No per-project

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 ## [Unreleased]
 
 ### Added
+- Make orch detached-by-default; add `--attach` (`20260427-orch-detach-default`) — 2026-04-28
+
 - Add missing gh-plan scripts to /apply-core install manifest (`20260324-apply-core-gh-scripts`) — 2026-03-24
 
 - Create GitHub Issue when plan is written, not when orch starts (`20260324-issue-on-plan-create`) — 2026-03-24

--- a/README.md
+++ b/README.md
@@ -537,6 +537,8 @@ The orchestrator runs execution plans with parallel worker agents in tmux. Each 
 bash ~/.degacore/scripts/orch-run.sh docs/exec-plans/active/20260302-add-auth-endpoint
 ```
 
+By default the orchestrator runs **detached**: the tmux session, engine, and workers all start in the background, and no terminal window or in-tmux ink dashboard is opened. Observe progress by reading `.orchestrator/plans/<slug>/state.json`, `heartbeat`, and the per-worker logs — or attach the canon-tui to those files. Pass `--attach` (alias `-a`) to opt back into the foreground experience: a terminal window attached to the tmux session plus the in-tmux `dashboard` window running the ink UI. The legacy `--background` flag is accepted as a silent no-op for one release (it requested what is now the default).
+
 ### Engine heartbeat
 
 The engine writes a Unix timestamp to `$ORCH_STATE_DIR/plans/$SLUG/heartbeat` at regular intervals — every poll cycle and before/after major steps (worker spawn, review, SHIP/FAIL). The dashboard and garbage collector use this file to detect whether the engine is alive or hung.

--- a/scripts/orch-run.sh
+++ b/scripts/orch-run.sh
@@ -6,19 +6,24 @@
 # The poll loop, worker spawning, review, and cleanup all run inside
 # orch-engine.sh in a tmux window named "engine".
 #
-# Usage: scripts/orch-run.sh <slug> [--issue N] [--max-workers N] [--max-iterations N] [--background]
+# Usage: scripts/orch-run.sh <slug> [--issue N] [--max-workers N] [--max-iterations N] [--attach|-a]
 #        scripts/orch-run.sh --gc [--dry-run]
 #
 # Options:
 #   --issue N            Fetch plan from GitHub Issue #N instead of local plan.md
 #   --max-workers N      Max concurrent workers (default: 4)
 #   --max-iterations N   Max review/rework iterations per item (default: 3)
-#   --background         Headless mode — tmux only, no display windows
+#   --attach, -a         Open a terminal attached to the tmux session and run
+#                        the in-tmux dashboard window (legacy foreground mode).
+#                        Default is detached: tmux session + engine only, no
+#                        terminal window opened, no dashboard window created.
+#   --background         Deprecated silent no-op (kept for one release; the
+#                        former --background behavior is now the default).
 #   --gc [--dry-run]     Run garbage collection on stale orch-* tmux sessions
 #
 # Example: scripts/orch-run.sh 20260309-orch-smoke-test
 # Example: scripts/orch-run.sh 20260309-orch-smoke-test --issue 42
-# Example: scripts/orch-run.sh 20260309-orch-smoke-test --background
+# Example: scripts/orch-run.sh 20260309-orch-smoke-test --attach
 
 set -euo pipefail
 
@@ -56,7 +61,7 @@ SLUG=""
 ISSUE_NUMBER=""
 MAX_WORKERS=4
 MAX_ITERATIONS=3
-BACKGROUND=false
+ATTACH=false
 
 while [[ $# -gt 0 ]]; do
   case "$1" in
@@ -80,13 +85,19 @@ while [[ $# -gt 0 ]]; do
     MAX_ITERATIONS="${2:-3}"
     shift 2
     ;;
+  --attach | -a)
+    ATTACH=true
+    shift
+    ;;
   --background)
-    BACKGROUND=true
+    # Deprecated: this used to opt out of the dashboard/display windows.
+    # That behavior is now the default, so the flag is a silent no-op
+    # for one release to keep existing callers working.
     shift
     ;;
   -*)
     echo "error: unknown option: $1" >&2
-    echo "usage: orch-run.sh <slug> [--issue N] [--max-workers N] [--max-iterations N] [--background] | --gc [--dry-run]" >&2
+    echo "usage: orch-run.sh <slug> [--issue N] [--max-workers N] [--max-iterations N] [--attach|-a] | --gc [--dry-run]" >&2
     exit 1
     ;;
   *)
@@ -97,7 +108,7 @@ while [[ $# -gt 0 ]]; do
 done
 
 if [[ -z "${SLUG}" ]]; then
-  echo "error: usage: orch-run.sh <slug> [--issue N] [--max-workers N] [--max-iterations N] [--background] | --gc [--dry-run]" >&2
+  echo "error: usage: orch-run.sh <slug> [--issue N] [--max-workers N] [--max-iterations N] [--attach|-a] | --gc [--dry-run]" >&2
   exit 1
 fi
 
@@ -367,16 +378,22 @@ fi
 orch_master_register "${SLUG}"
 orch_master_update_progress "${SLUG}"
 
-# --- Create tmux session with dashboard ---
+# --- Create tmux session ---
+#
+# In attach mode (--attach/-a) the session's first window is the
+# dashboard running the ink UI; the engine joins as a second window.
+# In the default detached mode there is no dashboard — the session is
+# created empty so the engine window can be attached to it below.
 
 if ! tmux has-session -t "${TMUX_SESSION}" 2>/dev/null; then
-  TERMINAL_UI_CLI="${SCRIPT_DIR}/terminal-ui/dist/cli.js"
-  HEARTBEAT_FILE="${ORCH_STATE_DIR}/plans/${SLUG}/heartbeat"
-  GRACE_TIMEOUT="${ORCH_DASHBOARD_TIMEOUT:-60}"
-  # Dashboard loop: restart node UI on crash, check engine liveness via
-  # heartbeat file staleness (>5min) and tmux window existence every 2s.
-  # When engine is dead, enter a grace period then kill the session.
-  read -r -d '' DASH_CMD <<-DASHEOF || true
+  if [[ "${ATTACH}" == true ]]; then
+    TERMINAL_UI_CLI="${SCRIPT_DIR}/terminal-ui/dist/cli.js"
+    HEARTBEAT_FILE="${ORCH_STATE_DIR}/plans/${SLUG}/heartbeat"
+    GRACE_TIMEOUT="${ORCH_DASHBOARD_TIMEOUT:-60}"
+    # Dashboard loop: restart node UI on crash, check engine liveness via
+    # heartbeat file staleness (>5min) and tmux window existence every 2s.
+    # When engine is dead, enter a grace period then kill the session.
+    read -r -d '' DASH_CMD <<-DASHEOF || true
 	grace_start=0
 	while true; do
 	  node '${TERMINAL_UI_CLI}' --orch '${ORCH_STATE_FILE}' 2>/dev/null
@@ -410,15 +427,15 @@ if ! tmux has-session -t "${TMUX_SESSION}" 2>/dev/null; then
 	  sleep 2
 	done
 	DASHEOF
-  tmux new-session -d -s "${TMUX_SESSION}" -n "dashboard" "${DASH_CMD}"
+    tmux new-session -d -s "${TMUX_SESSION}" -n "dashboard" "${DASH_CMD}"
+  else
+    tmux new-session -d -s "${TMUX_SESSION}"
+  fi
 fi
 
 # --- Start engine as a tmux window ---
 
 ENGINE_ARGS="${SLUG} --max-workers ${MAX_WORKERS} --max-iterations ${MAX_ITERATIONS}"
-if [[ "${BACKGROUND}" == true ]]; then
-  ENGINE_ARGS="${ENGINE_ARGS} --background"
-fi
 
 LOG_FILE=$(orch_plan_log_file "${SLUG}")
 orch_ensure_plan_dirs "${SLUG}"
@@ -426,9 +443,9 @@ orch_ensure_plan_dirs "${SLUG}"
 tmux new-window -d -t "${TMUX_SESSION}" -n "engine" \
   "cd '${REPO_ROOT}' && GH_SYNC='${GH_SYNC}' bash '${SCRIPT_DIR}/orch-engine.sh' ${ENGINE_ARGS} 2>&1 | tee '${LOG_FILE}'; echo '--- engine exited ---'; sleep 30; tmux kill-session -t '${TMUX_SESSION}' 2>/dev/null"
 
-# --- Open display windows (foreground mode) ---
+# --- Open display window (attach mode only) ---
 
-if [[ "${BACKGROUND}" == false ]]; then
+if [[ "${ATTACH}" == true ]]; then
   bash "${SCRIPT_DIR}/orch-display.sh" "${TMUX_SESSION}" || true
 fi
 

--- a/tests/orch/test_attach_default.bats
+++ b/tests/orch/test_attach_default.bats
@@ -1,0 +1,200 @@
+#!/usr/bin/env bats
+#
+# Tests for the detached-by-default behavior of scripts/orch-run.sh.
+#
+# The default invocation must run orchestration without a tmux
+# `dashboard` window and without invoking `orch-display.sh`. The
+# `--attach` flag (alias `-a`) opts back into the legacy foreground
+# behavior. `--background` is preserved as a silent no-op so existing
+# callers do not break.
+#
+# Strategy: spin up a self-contained git repo in TEST_TMP that mirrors
+# the real repo's `scripts/` tree. tmux, node, and orch-display.sh are
+# replaced by recording shims that append their argv to log files.
+# orch-run.sh is invoked against a stub plan; assertions inspect the
+# recorded calls. Because tmux is shimmed, no real session, engine, or
+# worker is ever spawned — the test exercises orch-run.sh's launch
+# logic only.
+
+REPO_ROOT_REAL="$(cd "$(dirname "${BATS_TEST_FILENAME}")/../.." && pwd)"
+
+setup() {
+  TEST_TMP="$(mktemp -d -t orch-attach-XXXXXX)"
+  export TEST_TMP
+
+  SLUG="attach-test"
+  export SLUG
+
+  # Mirror the real scripts/ tree so orch-run.sh's `source` and
+  # `${SCRIPT_DIR}/...` lookups resolve inside the test repo.
+  mkdir -p "${TEST_TMP}/scripts"
+  cp -R "${REPO_ROOT_REAL}/scripts/." "${TEST_TMP}/scripts/"
+
+  # Recording log files.
+  TMUX_LOG="${TEST_TMP}/tmux-calls.log"
+  NODE_LOG="${TEST_TMP}/node-calls.log"
+  DISPLAY_LOG="${TEST_TMP}/orch-display-calls.log"
+  export TMUX_LOG NODE_LOG DISPLAY_LOG
+  : >"${TMUX_LOG}"
+  : >"${NODE_LOG}"
+  : >"${DISPLAY_LOG}"
+
+  # Replace orch-display.sh in the copied tree with a recording stub.
+  # orch-run.sh invokes it as `bash "${SCRIPT_DIR}/orch-display.sh"`.
+  cat >"${TEST_TMP}/scripts/orch-display.sh" <<'STUB'
+#!/usr/bin/env bash
+printf '%s\n' "$*" >>"${DISPLAY_LOG}"
+exit 0
+STUB
+  chmod +x "${TEST_TMP}/scripts/orch-display.sh"
+
+  # PATH shims for tmux + node. Each records its argv and emulates the
+  # minimum behavior orch-run.sh expects.
+  SHIM_DIR="${TEST_TMP}/shims"
+  mkdir -p "${SHIM_DIR}"
+
+  cat >"${SHIM_DIR}/tmux" <<'SHIM'
+#!/usr/bin/env bash
+printf '%s\n' "$*" >>"${TMUX_LOG}"
+case "$1" in
+has-session) exit 1 ;;
+list-windows) exit 0 ;;
+*) exit 0 ;;
+esac
+SHIM
+  chmod +x "${SHIM_DIR}/tmux"
+
+  cat >"${SHIM_DIR}/node" <<'SHIM'
+#!/usr/bin/env bash
+printf '%s\n' "$*" >>"${NODE_LOG}"
+exit 0
+SHIM
+  chmod +x "${SHIM_DIR}/node"
+
+  export PATH="${SHIM_DIR}:${PATH}"
+
+  # Minimal repo state: github.sync false so the GH-issue auto-create
+  # path is skipped and the plan lives under docs/exec-plans/active/.
+  cat >"${TEST_TMP}/dega-core.yaml" <<'YAML'
+provider: github
+github:
+  sync: false
+YAML
+
+  PLAN_DIR="${TEST_TMP}/docs/exec-plans/active/${SLUG}"
+  mkdir -p "${PLAN_DIR}"
+  cat >"${PLAN_DIR}/plan.md" <<'PLAN'
+# Plan: attach-test
+
+## Progress log
+
+- [ ] do nothing
+PLAN
+
+  # Initialize a real git repo so `git rev-parse --show-toplevel` works
+  # and orch_create_worktree can `git worktree add` against HEAD. The
+  # plan must be committed to satisfy the dirty-plan guard.
+  git -C "${TEST_TMP}" init --quiet --initial-branch=main
+  git -C "${TEST_TMP}" config user.email "test@example.com"
+  git -C "${TEST_TMP}" config user.name "Test"
+  git -C "${TEST_TMP}" add -A
+  git -C "${TEST_TMP}" commit --quiet -m "init"
+
+  # Pin state + repo roots into the test repo for full isolation.
+  export ORCH_REPO_ROOT="${TEST_TMP}"
+  export ORCH_STATE_DIR="${TEST_TMP}/.orchestrator"
+
+  cd "${TEST_TMP}"
+}
+
+teardown() {
+  if [[ -n "${TEST_TMP:-}" && -d "${TEST_TMP}" ]]; then
+    rm -rf "${TEST_TMP}"
+  fi
+}
+
+# Run orch-run.sh from the test repo, capturing status + output.
+# Note: a leading `!` does not trigger bash errexit, so negative
+# assertions in this file always go through `run` + an explicit status
+# check rather than `! grep ...`.
+run_orch() {
+  run bash "${TEST_TMP}/scripts/orch-run.sh" "${SLUG}" "$@"
+}
+
+assert_orch_ok() {
+  if [ "${status}" -ne 0 ]; then
+    printf 'orch-run.sh failed (status=%s)\n--- output ---\n%s\n' \
+      "${status}" "${output}" >&2
+    return 1
+  fi
+}
+
+assert_grep() {
+  run grep -qE "$1" "$2"
+  [ "${status}" -eq 0 ]
+}
+
+refute_grep() {
+  run grep -qE "$1" "$2"
+  [ "${status}" -ne 0 ]
+}
+
+# --- (1) default = detached ---
+
+@test "default invocation creates session with no dashboard window and does not invoke orch-display.sh" {
+  run_orch
+  assert_orch_ok
+  assert_grep '^new-session ' "${TMUX_LOG}"
+  assert_grep '^new-window .* -n engine' "${TMUX_LOG}"
+  refute_grep '^new-session .* -n dashboard' "${TMUX_LOG}"
+  [ ! -s "${DISPLAY_LOG}" ]
+  # node is never executed in detached mode (no dashboard loop). The
+  # tmux shim records the new-session command string but does not
+  # execute it, so node only runs if orch-run.sh itself spawns it.
+  [ ! -s "${NODE_LOG}" ]
+}
+
+# --- (2) --attach = legacy foreground ---
+
+@test "--attach creates a dashboard window and invokes orch-display.sh" {
+  run_orch --attach
+  assert_orch_ok
+  assert_grep '^new-session .* -n dashboard' "${TMUX_LOG}"
+  assert_grep '^new-window .* -n engine' "${TMUX_LOG}"
+  [ -s "${DISPLAY_LOG}" ]
+}
+
+# --- (3) -a alias matches --attach ---
+
+@test "-a short alias matches --attach" {
+  run_orch -a
+  assert_orch_ok
+  assert_grep '^new-session .* -n dashboard' "${TMUX_LOG}"
+  [ -s "${DISPLAY_LOG}" ]
+}
+
+# --- (4) --background = silent no-op (matches default) ---
+
+@test "--background exits 0 silently and behaves identically to default" {
+  run_orch --background
+  assert_orch_ok
+  refute_grep '^new-session .* -n dashboard' "${TMUX_LOG}"
+  [ ! -s "${DISPLAY_LOG}" ]
+  # No "unknown option" or surprise error noise from the deprecation.
+  [[ "${output}" != *"unknown option"* ]]
+  [[ "${output}" != *"Unknown option"* ]]
+}
+
+# --- (5) detached session ends cleanly after engine exit ---
+#
+# In detached mode the dashboard window does not exist, so the engine
+# window's command string is the only thing keeping the session alive
+# after engine exit. It must include the `tmux kill-session` cleanup
+# epilogue so the session is reaped without operator intervention.
+
+@test "detached session has cleanup wired so it ends after engine exit" {
+  run_orch
+  assert_orch_ok
+  assert_grep '^new-window .* -n engine' "${TMUX_LOG}"
+  assert_grep '^new-window .* -n engine .*tmux kill-session' "${TMUX_LOG}"
+}

